kernel/src/engine/arrow_utils.rs

```diff
@@ -1,89 +1,895 @@
     } else {
         // requested an order different from the parquet, reorder
-        let input_schema = input_data.schema();
-        let mut fields = Vec::with_capacity(requested_ordering.len());
-        let reordered_columns = requested_ordering
-            .iter()
-            .map(|index| {
-                fields.push(input_schema.field(*index).clone());
-                input_data.column(*index).clone() // cheap Arc clone
-            })
+        debug!("Have requested reorder {requested_ordering:#?} on {input_data:?}");
+        let num_rows = input_data.len();
+        let (input_fields, mut input_cols, null_buffer) = input_data.into_parts();
+        let mut final_fields_cols: Vec<FieldArrayOpt> = std::iter::repeat_with(|| None)
+            .take(requested_ordering.len())
+            .collect();
+        for (parquet_position, reorder_index) in requested_ordering.iter().enumerate() {
+            // for each item, reorder_index.index() tells us where to put it, and its position in
+            // requested_ordering tells us where it is in the parquet data
+            match reorder_index {
+                ReorderIndex::Child { index, children } => {
+                    let mut placeholder: Arc<dyn ArrowArray> =
+                        Arc::new(StructArray::new_empty_fields(0, None));
+                    std::mem::swap(&mut input_cols[parquet_position], &mut placeholder);
+                    // placeholder now holds our struct array that we want to reorder
+                    let struct_array: StructArray = placeholder.into_data().into();
+                    let result_array = reorder_struct_array(struct_array, children)?;
+                    // create the new field specifying the correct order for the struct
+                    let new_field = Arc::new(ArrowField::new_struct(
+                        input_fields[parquet_position].name(),
+                        result_array.fields().clone(),
+                        input_fields[parquet_position].is_nullable(),
+                    ));
+                    final_fields_cols[*index] = Some((new_field, Arc::new(result_array)));
+                }
+                ReorderIndex::Index { index } => {
+                    final_fields_cols[*index] = Some((
+                        input_fields[parquet_position].clone(), // cheap Arc clone
+                        input_cols[parquet_position].clone(),   // cheap Arc clone
+                    ));
+                }
+                ReorderIndex::Null { index, field } => {
+                    let null_arry = Arc::new(new_null_array(field.data_type(), num_rows));
+                    let field = field.clone(); // cheap Arc clone
+                    final_fields_cols[*index] = Some((field, null_arry));
+                }
+            }
+        }
+        let fields = Fields::from_iter(
+            final_fields_cols
+                .iter()
+                .map(|fco| fco.as_ref().unwrap().0.clone()), // cheap Arc clone
```

Comment:
Could we maybe flatten instead of unwrap so we don't potentially panic on failure?

Reply:
Well, if it's `None` something has gone _very wrong_. We _need_ to have grabbed all the input columns and put them in their correct places. Anything left as `None` is a column we've accidentally dropped, so maybe `panic` is the right thing. 

I could change it to an `expect("Accidentally dropped a column when reordering, please report a bug against kernel")` if you think that would be better :)

Reply:
No more panic.  Have added the subtask.